File: android/app/src/main/java/com/example/HaliSahaApp/data/models/AppNotification.kt
package com.example.HaliSahaApp.data.models

import android.text.format.DateUtils
import com.google.firebase.firestore.DocumentId
import java.text.SimpleDateFormat
import java.util.*

// MARK: - AppNotification Model
data class AppNotification(
    @DocumentId
    val id: String? = null,
    val userId: String = "",               // Bildirimin gönderildiği kullanıcı
    val title: String = "",
    val body: String = "",
    val type: NotificationType = NotificationType.SYSTEM,
    val data: NotificationData? = null,      // Ek veriler (navigasyon için)
    val isRead: Boolean = false,
    val createdAt: Date = Date()
) {
    // MARK: - Computed Properties
    val icon: String get() = type.icon
    val color: String get() = type.color

    val relativeTime: String
        get() = DateUtils.getRelativeTimeSpanString(
            createdAt.time,
            System.currentTimeMillis(),
            DateUtils.MINUTE_IN_MILLIS,
            DateUtils.FORMAT_ABBREV_RELATIVE
        ).toString()

    val formattedDate: String
        get() {
            return when {
                DateUtils.isToday(createdAt.time) -> {
                    SimpleDateFormat("'Bugün' HH:mm", Locale("tr", "TR")).format(createdAt)
                }
                isYesterday(createdAt) -> {
                    SimpleDateFormat("'Dün' HH:mm", Locale("tr", "TR")).format(createdAt)
                }
                else -> {
                    SimpleDateFormat("d MMM, HH:mm", Locale("tr", "TR")).format(createdAt)
                }
            }
        }

    private fun isYesterday(date: Date): Boolean {
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.DAY_OF_YEAR, -1)
        val yesterday = calendar.time
        val fmt = SimpleDateFormat("yyyyMMdd", Locale.getDefault())
        return fmt.format(date) == fmt.format(yesterday)
    }

    // MARK: - Factory Methods
    companion object {
        fun bookingConfirmed(userId: String, booking: Booking) = AppNotification(
            userId = userId,
            title = "Rezervasyon Onaylandı ✅",
            body = "${booking.facilityName} - ${booking.pitchName} için ${booking.formattedDate} tarihli rezervasyonunuz onaylandı.",
            type = NotificationType.BOOKING_CONFIRMED,
            data = NotificationData(bookingId = booking.id, facilityId = booking.facilityId)
        )

        fun bookingCancelled(userId: String, booking: Booking, reason: String? = null): AppNotification {
            val body = if (!reason.isNullOrEmpty()) {
                "${booking.facilityName} – ${booking.formattedDate} tarihli rezervasyonunuz iptal edildi. Sebep: $reason"
            } else {
                "${booking.facilityName} – ${booking.formattedDate} tarihli rezervasyonunuz iptal edildi."
            }
            return AppNotification(
                userId = userId,
                title = "Rezervasyon İptali ❌",
                body = body,
                type = NotificationType.BOOKING_CANCELLED,
                data = NotificationData(bookingId = booking.id, facilityId = booking.facilityId)
            )
        }

        fun newBookingForAdmin(adminId: String, booking: Booking): AppNotification {
            val actionText = if (booking.status == BookingStatus.pending) {
                "için ödeme yaptı ve onayınızı bekliyor."
            } else {
                "için ayırdı."
            }
            return AppNotification(
                userId = adminId,
                title = "Yeni Rezervasyon 🎫",
                body = "${booking.userFullName} – ${booking.pitchName} sahasını ${booking.formattedDate} (${booking.timeSlotString}) $actionText",
                type = NotificationType.BOOKING_CONFIRMED,
                data = NotificationData(bookingId = booking.id, facilityId = booking.facilityId, pitchId = booking.pitchId)
            )
        }

        fun bookingCancelledByUser(adminId: String, booking: Booking) = AppNotification(
            userId = adminId,
            title = "Rezervasyon İptal Edildi ❌",
            body = "${booking.userFullName} – ${booking.pitchName} için ${booking.formattedDate} tarihli rezervasyonunu iptal etti.",
            type = NotificationType.BOOKING_CANCELLED,
            data = NotificationData(bookingId = booking.id, facilityId = booking.facilityId, pitchId = booking.pitchId)
        )

        fun reviewReceived(adminId: String, facilityName: String, review: Review): AppNotification {
            val stars = String.format(Locale.US, "%.1f", review.overallRating)
            val snippet = if (!review.comment.isNullOrEmpty()) {
                val truncated = if (review.comment.length > 100) "${review.comment.take(100)}…" else review.comment
                "${review.userName} ($stars★): \"$truncated\""
            } else {
                "${review.userName} tesisini $stars yıldızla değerlendirdi."
            }
            return AppNotification(
                userId = adminId,
                title = "Yeni Değerlendirme ⭐️",
                body = snippet,
                type = NotificationType.REVIEW_RECEIVED,
                data = NotificationData(facilityId = review.facilityId, pitchId = review.pitchId, reviewId = review.id)
            )
        }

        fun bookingReminder(userId: String, booking: Booking, hoursLeft: Int) = AppNotification(
            userId = userId,
            title = "Maç Hatırlatması ⚽",
            body = "${booking.facilityName} - ${booking.pitchName}'da maçınıza $hoursLeft saat kaldı!",
            type = NotificationType.BOOKING_REMINDER,
            data = NotificationData(bookingId = booking.id, facilityId = booking.facilityId)
        )

        fun matchInvite(userId: String, senderName: String, groupId: String, bookingId: String) = AppNotification(
            userId = userId,
            title = "Maç Daveti 🎯",
            body = "$senderName sizi bir maça davet etti.",
            type = NotificationType.MATCH_INVITE,
            data = NotificationData(groupId = groupId, bookingId = bookingId)
        )

        fun joinRequestReceived(userId: String, applicantName: String, postId: String) = AppNotification(
            userId = userId,
            title = "Katılma İsteği 🙋",
            body = "$applicantName maçınıza katılmak istiyor.",
            type = NotificationType.JOIN_REQUEST,
            data = NotificationData(matchPostId = postId)
        )

        fun joinRequestAccepted(userId: String, facilityName: String, groupId: String) = AppNotification(
            userId = userId,
            title = "İstek Kabul Edildi 🎉",
            body = "$facilityName'daki maça katılma isteğiniz kabul edildi!",
            type = NotificationType.JOIN_REQUEST_ACCEPTED,
            data = NotificationData(groupId = groupId)
        )

        fun newMessage(userId: String, senderName: String, groupId: String, groupName: String) = AppNotification(
            userId = userId,
            title = groupName,
            body = "$senderName bir mesaj gönderdi.",
            type = NotificationType.NEW_MESSAGE,
            data = NotificationData(groupId = groupId)
        )

        fun newFollower(userId: String, followerName: String, followerId: String) = AppNotification(
            userId = userId,
            title = "Yeni Takipçi 👤",
            body = "$followerName sizi takip etmeye başladı.",
            type = NotificationType.NEW_FOLLOWER,
            data = NotificationData(userId = followerId)
        )
    }
}

// MARK: - Notification Type Enum
enum class NotificationType(val rawValue: String, val icon: String, val color: String) {
    BOOKING_CONFIRMED("bookingConfirmed", "check_circle", "green"),
    BOOKING_CANCELLED("bookingCancelled", "cancel", "red"),
    BOOKING_REMINDER("bookingReminder", "notifications_active", "orange"),
    MATCH_INVITE("matchInvite", "sports_soccer", "blue"),
    JOIN_REQUEST("joinRequest", "person_add", "blue"),
    JOIN_REQUEST_ACCEPTED("joinRequestAccepted", "verified", "green"),
    JOIN_REQUEST_REJECTED("joinRequestRejected", "block", "red"),
    NEW_MESSAGE("newMessage", "chat", "purple"),
    NEW_FOLLOWER("newFollower", "person_add_alt", "pink"),
    FACILITY_APPROVED("facilityApproved", "business", "green"),
    FACILITY_REJECTED("facilityRejected", "domain_disabled", "red"),
    REVIEW_RECEIVED("reviewReceived", "grade", "yellow"),
    PROMOTIONAL("promotional", "redeem", "indigo"),
    SYSTEM("system", "notifications", "gray");

    val category: NotificationCategory
        get() = when (this) {
            BOOKING_CONFIRMED, BOOKING_CANCELLED, BOOKING_REMINDER -> NotificationCategory.BOOKING
            MATCH_INVITE, JOIN_REQUEST, JOIN_REQUEST_ACCEPTED, JOIN_REQUEST_REJECTED, NEW_MESSAGE, NEW_FOLLOWER -> NotificationCategory.SOCIAL
            else -> NotificationCategory.SYSTEM
        }
}

// MARK: - Notification Category Enum
enum class NotificationCategory(val rawValue: String, val displayName: String) {
    BOOKING("booking", "Rezervasyonlar"),
    SOCIAL("social", "Sosyal"),
    SYSTEM("system", "Sistem");
}

// MARK: - Notification Data
data class NotificationData(
    val bookingId: String? = null,
    val facilityId: String? = null,
    val pitchId: String? = null,
    val groupId: String? = null,
    val matchPostId: String? = null,
    val userId: String? = null,
    val reviewId: String? = null
)

File: ios/HaliSahaApp/HaliSahaApp/Models/AppNotification.swift
//
//  AppNotification.swift
//  HaliSahaApp
//
//  Uygulama içi bildirim veri modeli
//
//  Created by Mehmet Mert Mazıcı on 23.12.2025.
//


import Foundation
import FirebaseFirestore

// MARK: - AppNotification Model
struct AppNotification: Identifiable, Codable, Hashable {
    @DocumentID var id: String?
    var userId: String               // Bildirimin gönderildiği kullanıcı
    var title: String
    var body: String
    var type: NotificationType
    var data: NotificationData?      // Ek veriler (navigasyon için)
    var isRead: Bool
    var createdAt: Date
    
    // MARK: - Computed Properties
    var icon: String {
        type.icon
    }
    
    var color: String {
        type.color
    }
    
    var relativeTime: String {
        let formatter = RelativeDateTimeFormatter()
        formatter.locale = Locale(identifier: "tr_TR")
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: createdAt, relativeTo: Date())
    }
    
    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "tr_TR")
        
        if Calendar.current.isDateInToday(createdAt) {
            formatter.dateFormat = "'Bugün' HH:mm"
        } else if Calendar.current.isDateInYesterday(createdAt) {
            formatter.dateFormat = "'Dün' HH:mm"
        } else {
            formatter.dateFormat = "d MMM, HH:mm"
        }
        return formatter.string(from: createdAt)
    }
    
    // MARK: - Initializer
    init(
        id: String? = nil,
        userId: String,
        title: String,
        body: String,
        type: NotificationType,
        data: NotificationData? = nil,
        isRead: Bool = false,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.userId = userId
        self.title = title
        self.body = body
        self.type = type
        self.data = data
        self.isRead = isRead
        self.createdAt = createdAt
    }
    
    // MARK: - Factory Methods
    static func bookingConfirmed(userId: String, booking: Booking) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "Rezervasyon Onaylandı ✅",
            body: "\(booking.facilityName) - \(booking.pitchName) için \(booking.formattedDate) tarihli rezervasyonunuz onaylandı.",
            type: .bookingConfirmed,
            data: NotificationData(bookingId: booking.id, facilityId: booking.facilityId)
        )
    }
    
    static func bookingReminder(userId: String, booking: Booking, hoursLeft: Int) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "Maç Hatırlatması ⚽",
            body: "\(booking.facilityName) - \(booking.pitchName)'da maçınıza \(hoursLeft) saat kaldı!",
            type: .bookingReminder,
            data: NotificationData(bookingId: booking.id, facilityId: booking.facilityId)
        )
    }

    static func bookingCancelled(userId: String, booking: Booking, reason: String? = nil) -> AppNotification {
        let body: String
        if let reason, !reason.isEmpty {
            body = "\(booking.facilityName) – \(booking.formattedDate) tarihli rezervasyonunuz iptal edildi. Sebep: \(reason)"
        } else {
            body = "\(booking.facilityName) – \(booking.formattedDate) tarihli rezervasyonunuz iptal edildi."
        }
        return AppNotification(
            userId: userId,
            title: "Rezervasyon İptali ❌",
            body: body,
            type: .bookingCancelled,
            data: NotificationData(bookingId: booking.id, facilityId: booking.facilityId)
        )
    }

    /// Admin'e: yeni rezervasyon (kullanıcı oluşturduğunda)
    static func newBookingForAdmin(adminId: String, booking: Booking) -> AppNotification {
        let actionText =
            booking.status == .pending
            ? "için ödeme yaptı ve onayınızı bekliyor."
            : "için ayırdı."

        return AppNotification(
            userId: adminId,
            title: "Yeni Rezervasyon 🎫",
            body: "\(booking.userFullName) – \(booking.pitchName) sahasını \(booking.formattedDate) (\(booking.timeSlotString)) \(actionText)",
            type: .bookingConfirmed,
            data: NotificationData(bookingId: booking.id, facilityId: booking.facilityId, pitchId: booking.pitchId)
        )
    }

    /// Admin'e: kullanıcı rezervasyonunu iptal etti
    static func bookingCancelledByUser(adminId: String, booking: Booking) -> AppNotification {
        AppNotification(
            userId: adminId,
            title: "Rezervasyon İptal Edildi ❌",
            body: "\(booking.userFullName) – \(booking.pitchName) için \(booking.formattedDate) tarihli rezervasyonunu iptal etti.",
            type: .bookingCancelled,
            data: NotificationData(bookingId: booking.id, facilityId: booking.facilityId, pitchId: booking.pitchId)
        )
    }

    /// Admin'e: yeni değerlendirme alındı
    static func reviewReceived(
        adminId: String,
        facilityName: String,
        review: Review
    ) -> AppNotification {
        let stars = String(format: "%.1f", review.overallRating)
        let snippet: String = {
            guard let comment = review.comment, !comment.isEmpty else {
                return "\(review.userName) tesisini \(stars) yıldızla değerlendirdi."
            }
            let truncated = comment.count > 100
                ? String(comment.prefix(100)) + "…"
                : comment
            return "\(review.userName) (\(stars)★): \"\(truncated)\""
        }()

        return AppNotification(
            userId: adminId,
            title: "Yeni Değerlendirme ⭐️",
            body: snippet,
            type: .reviewReceived,
            data: NotificationData(
                facilityId: review.facilityId,
                pitchId: review.pitchId,
                reviewId: review.id
            )
        )
    }
    
    static func matchInvite(userId: String, senderName: String, groupId: String, bookingId: String) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "Maç Daveti 🎯",
            body: "\(senderName) sizi bir maça davet etti.",
            type: .matchInvite,
            data: NotificationData(bookingId: bookingId, groupId: groupId)
        )
    }
    
    static func joinRequestReceived(userId: String, applicantName: String, postId: String) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "Katılma İsteği 🙋",
            body: "\(applicantName) maçınıza katılmak istiyor.",
            type: .joinRequest,
            data: NotificationData(matchPostId: postId)
        )
    }
    
    static func joinRequestAccepted(userId: String, facilityName: String, groupId: String) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "İstek Kabul Edildi 🎉",
            body: "\(facilityName)'daki maça katılma isteğiniz kabul edildi!",
            type: .joinRequestAccepted,
            data: NotificationData(groupId: groupId)
        )
    }
    
    static func newMessage(userId: String, senderName: String, groupId: String, groupName: String) -> AppNotification {
        AppNotification(
            userId: userId,
            title: groupName,
            body: "\(senderName) bir mesaj gönderdi.",
            type: .newMessage,
            data: NotificationData(groupId: groupId)
        )
    }
    
    static func newFollower(userId: String, followerName: String, followerId: String) -> AppNotification {
        AppNotification(
            userId: userId,
            title: "Yeni Takipçi 👤",
            body: "\(followerName) sizi takip etmeye başladı.",
            type: .newFollower,
            data: NotificationData(userId: followerId)
        )
    }
}

// MARK: - Notification Type
enum NotificationType: String, Codable, CaseIterable {
    // Rezervasyon bildirimleri
    case bookingConfirmed = "bookingConfirmed"
    case bookingCancelled = "bookingCancelled"
    case bookingReminder = "bookingReminder"
    
    // Sosyal bildirimler
    case matchInvite = "matchInvite"
    case joinRequest = "joinRequest"
    case joinRequestAccepted = "joinRequestAccepted"
    case joinRequestRejected = "joinRequestRejected"
    case newMessage = "newMessage"
    case newFollower = "newFollower"
    
    // Sistem bildirimleri
    case facilityApproved = "facilityApproved"
    case facilityRejected = "facilityRejected"
    case reviewReceived = "reviewReceived"
    case promotional = "promotional"
    case system = "system"
    
    var icon: String {
        switch self {
        case .bookingConfirmed: return "checkmark.circle.fill"
        case .bookingCancelled: return "xmark.circle.fill"
        case .bookingReminder: return "alarm.fill"
        case .matchInvite: return "sportscourt.fill"
        case .joinRequest: return "person.badge.plus"
        case .joinRequestAccepted: return "checkmark.seal.fill"
        case .joinRequestRejected: return "xmark.seal.fill"
        case .newMessage: return "message.fill"
        case .newFollower: return "person.fill.badge.plus"
        case .facilityApproved: return "building.2.fill"
        case .facilityRejected: return "building.2"
        case .reviewReceived: return "star.fill"
        case .promotional: return "gift.fill"
        case .system: return "bell.fill"
        }
    }
    
    var color: String {
        switch self {
        case .bookingConfirmed, .joinRequestAccepted, .facilityApproved:
            return "green"
        case .bookingCancelled, .joinRequestRejected, .facilityRejected:
            return "red"
        case .bookingReminder:
            return "orange"
        case .matchInvite, .joinRequest:
            return "blue"
        case .newMessage:
            return "purple"
        case .newFollower:
            return "pink"
        case .reviewReceived:
            return "yellow"
        case .promotional:
            return "indigo"
        case .system:
            return "gray"
        }
    }
    
    var category: NotificationCategory {
        switch self {
        case .bookingConfirmed, .bookingCancelled, .bookingReminder:
            return .booking
        case .matchInvite, .joinRequest, .joinRequestAccepted, .joinRequestRejected, .newMessage, .newFollower:
            return .social
        case .facilityApproved, .facilityRejected, .reviewReceived, .promotional, .system:
            return .system
        }
    }
}

// MARK: - Notification Category
enum NotificationCategory: String, CaseIterable {
    case booking = "booking"
    case social = "social"
    case system = "system"
    
    var displayName: String {
        switch self {
        case .booking: return "Rezervasyonlar"
        case .social: return "Sosyal"
        case .system: return "Sistem"
        }
    }
}

// MARK: - Notification Data (Navigasyon için ek veriler)
struct NotificationData: Codable, Hashable {
    var bookingId: String?
    var facilityId: String?
    var pitchId: String?
    var groupId: String?
    var matchPostId: String?
    var userId: String?
    var reviewId: String?
    
    init(
        bookingId: String? = nil,
        facilityId: String? = nil,
        pitchId: String? = nil,
        groupId: String? = nil,
        matchPostId: String? = nil,
        userId: String? = nil,
        reviewId: String? = nil
    ) {
        self.bookingId = bookingId
        self.facilityId = facilityId
        self.pitchId = pitchId
        self.groupId = groupId
        self.matchPostId = matchPostId
        self.userId = userId
        self.reviewId = reviewId
    }
}

// MARK: - Mock Data
extension AppNotification {
    static let mockNotifications: [AppNotification] = [
        AppNotification(
            id: "notif1",
            userId: "user123",
            title: "Rezervasyon Onaylandı ✅",
            body: "Yıldız Spor Tesisleri - Saha A için 28 Aralık tarihli rezervasyonunuz onaylandı.",
            type: .bookingConfirmed,
            data: NotificationData(bookingId: "booking123", facilityId: "facility123"),
            createdAt: Date().addingTimeInterval(-3600)
        ),
        AppNotification(
            id: "notif2",
            userId: "user123",
            title: "Maç Daveti 🎯",
            body: "Mehmet sizi bir maça davet etti.",
            type: .matchInvite,
            data: NotificationData(groupId: "group123"),
            createdAt: Date().addingTimeInterval(-7200)
        ),
        AppNotification(
            id: "notif3",
            userId: "user123",
            title: "Yeni Takipçi 👤",
            body: "Ali sizi takip etmeye başladı.",
            type: .newFollower,
            data: NotificationData(userId: "user789"),
            isRead: true,
            createdAt: Date().addingTimeInterval(-86400)
        ),
        AppNotification(
            id: "notif4",
            userId: "user123",
            title: "Maç Hatırlatması ⚽",
            body: "Yıldız Spor Tesisleri - Saha A'da maçınıza 2 saat kaldı!",
            type: .bookingReminder,
            data: NotificationData(bookingId: "booking123"),
            createdAt: Date().addingTimeInterval(-1800)
        )
    ]
}
